### PR TITLE
[SGen] Optimize thread pool enqueue to handle job batches more efficiently.

### PR DESCRIPTION
**Pull request by lateralusX**
When running parallel minor GC the scan of major and LOS is split up in several jobs based on number of available cores * 4 * 2. On an 8 core machine that will generate 32 major heap and 32 LOS scanning jobs. Current implementation queued each individual job into the thread pool but since the threads where not allowed to work on the items at this point, doing it
this way creates a lot of signalled threads + contention over shared mutex for each queued job.

Commit optimize this pattern adding the ability to add a batch of allocated jobs into the thread pool at the same time, signalling all threads when all, jobs have been added to the queue reducing the number of mutex acquire/release and waking all thread pool threads from 65 (including scan wbroots) down to 1.

This is an upstream of a change running in downstream repro for a little over a year giving a good performance boost for those platforms. As part of upstreaming I also did some performance benchmark around performance of this PR on desktop 8 core Windows PC. Test was primarily stressing the job dispatch as part of minor GC using a major heap of ~2.5 GB and a LOB heap of ~1 GB, doing around 1500 minor GC measures points per configuration (over stable GC state). The following is a summary of minor GC pause times comparing default, default with minor=simple-par and minor=simple-par + this PR:

  | Default | Simple-par | Simple-par + enqueue   optimization | Improvement | Improvement default Simple-par
-- | -- | -- | -- | -- | --
Stdev (µs) | 1143 | 1047 | 952 |   |  
Avg (µs) | 2585 | 2206 | 1629 | 36,98% | 26,17%
  |   |   |   |   |  
TrimMean 10% (µs) | 2548 | 2146 | 1576 | 38,17% | 26,58%
TrimMean 25% (µs) | 2535 | 2106 | 1548 | 38,94% | 26,50%
First quartile – 25th percentile (µs) | 1598 | 1295 | 965 | 39,63% | 25,54%
Second quartile – 50th percentile (µs) | 2449 | 2062 | 1515 | 38,16% | 26,55%
Third quartile – 75th percentile (µs) | 3588 | 2892 | 2116 | 41,04% | 26,84%

So to summarize, above indicates minor GC pause times improved with ~38% when running with the enqueue optimizations implemented in this PR on a desktop 8 core machine on a 2.5 GB major heap and 1 GB LOB heap. Results can off course vary depending on hardware, workload and platform, but benchmark above at least gives a good indication that batching jobs and reduce number of unnecessary kernel calls and pressure on scheduler have measurable positive improvements on minor GC pause times.

### CODE REVIEW

**Comment by BrzVlad** [comment]
I think this change is a bit ugly. I don't really like the way we allocate the job array through the workers api and the way it is used, the fact that we need to know its size and we allocate and deallocate it with each collection. I think all this functionality should be provided directly by the workers api and a simpler design could be as follows:

```
// If deferred, instead of sending the job to the threadpool, we can store it inside a static thread pool
// job array. If the array is at full capacity, flush all jobs to the threadpool. This array can be allocated
// when the worker contexts are created, with a static reasonable size that is optimal for our cases.
sgen_workers_enqueue_job (gboolean deferred);
// Enqueues to the threadpool all the jobs that accumulated in the deferred array. Called once by GC
// once all job allocation has finished.
sgen_workers_flush_deferred_jobs ();
```
How does this sound ?

**Comment by lateralusX** [comment]
We could do it that way, but what what I wanted to avoid was the need to take the mutex lock for each enqueued job, that's why you could add an array of jobs, so it would be similar to do all in one transaction. If the mutex lock is a spin lock and I believe there won't be any contention, then its probably not noticeable, but on platforms where this is not true (I know a couple of platforms), there will be a lot of unnecessary kernel calls just adding the things into the queue.

Could an option be for client to use a growable data structure, like the sgen_pointer_queue, then the client can then add all jobs into that and send it over to enqueue API in one transaction (just taking the mutex lock once when adding all jobs into).

I agree that we could move the signal into a deferred queue/context that can be flushed when full or explicit using API. If we just size that queue correctly (based on our number of work items API we use to calculate number of items) we should be able to queue up all what's needed and only get one signal waking up all workers when there is real work to be done.

Thoughs?

**Comment by lateralusX** [comment]
So based on some offline discussion, the idea was to not lock the deferred queue since it will only be handled by GC thread, so we if we do it that way and we add a BIG disclaimer on the API, there won't be any need to lock when adding into the deferred queue. I will work on change next week.

**Comment by lateralusX** [comment]
Implemented alternative API based on discussion, putting deferred queue besides job queue. It is pre-allocated to hold jobs generated during a minor GC without forcing unnecessary flushes into job queue (reduce number of mutex locks).

Since the deferred queue assumes GC lock held (since it won't take any explicit locks), it comes with a big disclaimer + its own enqueue and flush methods to make that assumption more clear than cloaking it in existing enqueue API.

Most of the implementation behaves the same as the original, batching up jobs in the deferred queue and flush it into job queue (but without signal workers). Just before dispatching all, it will do a last flush and signal, making sure all deferred jobs are in job queue and signal all workers.

Preliminary measures are inline with previous measures, but will do a new set of measures on both default as well as this new API and include a new table for reference.

**Comment by lateralusX** [comment]
New run of 1500 measures per configuration. This just compares default with minor=simple-par including changes in this PR to improve enqueue efficiency on job batching. Improvement in this measure is a little lower compared to initial and has less spread of values, still seeing ~30% perf improvement using simple-par (including this change). The tests primarily runs minor GC on stable heap, so work should be fairly constant between measures. Increasing the work done per dispatched job also showed even more performance boost on some platforms, but all depends on hardware, OS and work loads.

Test starts up by allocating a major heap of 1 GB and a 1 GB LOB including random sized byte arrays. It then does a full GC and starts a loop of 500 measure points, measuring "wall clock" per GC.Collect(0) using high res timer print result and sleep for 1 second per iteration. Test was run a 3 separate occasions per configuration adding up a total of 1500 measures per configuration.

  | Default | Simple-par + enqueue   optimization | Improvement
-- | -- | -- | --
Stdev (µs) | 920 | 1029 |   |
Avg (µs) | 1968 | 1428 | 27,46%
  |   |   |   
TrimMean 10% (µs) | 1898 | 1362 | 28,25%
TrimMean 25% (µs) | 1848 | 1332 | 27,91%
First quartile – 25th percentile (µs) | 1280 |  944 | 26,24%
Second quartile – 50th percentile (µs) | 1860 | 1321 | 28,96%
Third quartile – 75th percentile (µs) | 2379 | 1707 | 28,25%

File: mono/sgen/sgen-gc.c [hunk 1]
```diff
@@ -1901,6 +1906,8 @@ collect_nursery (const char *reason, gboolean is_overflow)
 
 	enqueue_scan_from_roots_jobs (&gc_thread_gray_queue, sgen_nursery_section->data, sgen_nursery_section->end_data, is_parallel ? NULL : object_ops_nopar, is_parallel);
 
+	sgen_workers_flush_deferred_jobs (GENERATION_NURSERY, is_parallel);
```
Comment (BrzVlad) [inline mono/sgen/sgen-gc.c:1909]: Is there a reason for flushing twice, since we already do it in enqueue_scan_from_roots_jobs ?

Comment (lateralusX) [reply mono/sgen/sgen-gc.c:1909]: It flushed but won't signal and then we just signal, but agree, due to the way the calls ended up, we could just flush and signal directly in enqueue_scan_from_roots_jobs. The intent was that enqueue_scan_from_roots_jobs might not be the last function called before dispatching workers, so it flushed since it generates a lot of jobs, but that might be a little over ambitious since we can rely on the auto flush (since I changed it to not signal, just flush). I will correct this, only do explicit flush+signal just before we dispatch workers.

**Comment by lateralusX** [comment]
Dropped explicit calls to flush in enqueue_scan_remembered_set_jobs and enqueue_scan_from_roots_jobs as well as update second table with new test data + added the default simple-par into the table as well.

File: mono/sgen/sgen-gc.c (hunk 1 above)
Comment (lateralusX) [reply mono/sgen/sgen-gc.c:1905]: Fixed.

**Comment by lateralusX** [comment]
@monojenkins rebuild failed

**Review by BrzVlad — APPROVED** [review]
Great numbers. We might want to consider using the parallel mode in the future for a wider set of platforms.
